README.md

```diff
@@ -4,128 +4,336 @@ Daniel Ehrenberg, Igalia. Stage 3
+5n / 2n
+// ↪ 2n
 
-Many (all?) other dynamically typed programming languages which have multiple numeric types implement a *numeric tower*. This forms an ordering between types--on the built-in numeric types, when an operator is used with operands from two types, the greater type is chosen as the domain, and the "less general" operand is cast to the "more general" type. Unfortunately, as the previous example shows, there is no "more general" type between arbitrary integers and double-precision floats. The typical resolution, then, is to take floats as the "more general" type.
+Number(151851850485185185047n)
+// ↪ 151851850485185200000
 
-Silently losing precision sometimes may be a problem, but in most dynamically typed programming languages which provide integers and floats, integers are written like `1` and floats are written like `1.0`. It's possible to scan code for operations which may introduce floating point precision by looking for a decimal point. JavaScript exacerbates the scope of losing precision by making the unfortunate decision that a simple literal like `1` is a float. So, if mixed-precision were allowed, an innocent calculation such as `2n ** 53n + 1` would produce the float `2**53`--defeating the core functionality of this feature.
+```
 
-To avoid this problem, this proposal bans implicit coercions between Numbers and BigInts, including operations which are mixed type. `1n + 1` throws a TypeError. So does passing `1n` as an argument into any JavaScript standard library function or Web API which expects a Number. Instead, to convert between types, an explicit call to `Number()` or `BigInt()` needs to be made to decide which domain to operate in. `0 === 0n` returns `false`.
+### Other Exceptions
 
-Comparisons form an exception to this rule: It's mathematically well-defined to allow comparison operators such as `<` and `==` compare between Numbers and BigInts. Unlike operators like `+`, there is no loss of precision, since the output is just a Boolean. Further, `==` is extended in comparisons with strings in analogous ways, such as `0n == ""` is `true`. Although we may not be able to extend this support to user-defined types, it seems sufficiently important for meeting user expectations that this proposal adds them as a one-off.
+Attempting to convert a fractional value to a `BigInt` throws an exception both when the value is represented as an `Number` and a `String`.
 
-## Design goals
+```js
+BigInt(1.5)
+// ↪ RangeError: The number 1.5 is not a safe integer and thus cannot be converted to a BigInt
 
-### Don't break user intuition
+BigInt('1.5')
+// ↪ SyntaxError: Cannot convert 1.5 to a BigInt
+
+```
+
+`Math.fround` and `|` throw exceptions when called with a `BigInt` to preserve the current expectations of `asm.js`.
```

Comment:
At this point, the asm.js motivation is fairly secondary. The main principle here is about not mixing BigInt and Number (these cases follow that pattern), as you documented on line 60. For some references to build the case for that: I gave a motivating example in the [JSConfEU talk slides](https://docs.google.com/presentation/d/1apPbAiv_-mJF35P31IjaII8UA6TwSynCA_zhfDEmgOE/edit#slide=id.g385aa41f00_0_218). This leads to [my best practices recommendation](https://docs.google.com/presentation/d/1apPbAiv_-mJF35P31IjaII8UA6TwSynCA_zhfDEmgOE/edit#slide=id.g3b79f6de99_0_49), which may make sense to reproduce here.

Reply:
That's a good idea. I'm not sure if it is too complex to note that these two operations are specc'd to return `Number`s only, as opposed to other operations which don't run into this restriction, but it should probably be fine. It's more important, I think, to let people know all the possible error cases up front, so they don't find themselves hunting for a possibility.